### PR TITLE
Address pandas FutureWarnings in test suite

### LINKED ISSUE #1892: Various pandas deprecation warnings from test suite
We've started getting several types of deprecation warnings from the tests:

```
pvlib/tests/test_modelchain.py::test_aoi_model_interp
  /home/runner/work/pvlib-python/pvlib-python/pvlib/tests/test_modelchain.py:1476: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`
    assert mc.results.ac[1] < 1
```

```
pvlib/tests/test_clearsky.py::test_detect_clearsky_window
  /home/runner/work/pvlib-python/pvlib-python/pvlib/tests/test_clearsky.py:634: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise in a future error of pandas. Value 'True' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.
    expected.iloc[-3:] = True
```

```
pvlib/tests/test_singlediode.py: 14 warnings
  /home/runner/work/pvlib-python/pvlib-python/pvlib/tests/test_singlediode.py:118: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
    joined[is_number] = joined[is_number].applymap(np.float64)
```

```
  /home/runner/work/pvlib-python/pvlib-python/pvlib/scaling.py:279: FutureWarning: DataFrame.fillna with 'method' is deprecated and will raise in a future version. Use obj.ffill() or obj.bfill() instead.
    df = df.fillna(method='bfill').fillna(method='ffill')
```

Here's an example run: https://github.com/pvlib/pvlib-python/actions/runs/6550206298/job/17788805508?pr=1827#step:9:65

Addressing these would be good first PRs for anyone wanting to get started with contributing to pvlib's codebase!  If you plan to submit a PR for one of these warnings, please comment here so that someone else does not duplicate the work.

### PR DESCRIPTION
**Pull request by matsuobasho**
- [ ] Closes #1892 
 - [X] I am familiar with the [contributing guidelines](https://pvlib-python.readthedocs.io/en/latest/contributing.html)
 - [ ] Tests added - NA since issue is about getting rid of warnings in tests
 - [ ] Updates entries in [`docs/sphinx/source/reference`](https://github.com/pvlib/pvlib-python/blob/main/docs/sphinx/source/reference) for API changes. - NA see above
 - [x] Adds description and name entries in the appropriate "what's new" file in [`docs/sphinx/source/whatsnew`](https://github.com/pvlib/pvlib-python/tree/main/docs/sphinx/source/whatsnew) for all changes. Includes link to the GitHub Issue with `` :issue:`num` `` or this Pull Request with `` :pull:`num` ``. Includes contributor name and/or GitHub username (link with `` :ghuser:`user` ``).  - NA
 - [ ] New code is fully documented. Includes [numpydoc](https://numpydoc.readthedocs.io/en/latest/format.html) compliant docstrings, examples, and comments where necessary.  NA
 - [X] Pull request is nearly complete and ready for detailed review.

Updated tests to get rid of warnings.

### CODE REVIEW

File: pvlib/tests/test_singlediode.py [hunk 1]
```diff
@@ -110,12 +110,12 @@ def build_precise_iv_curve_dataframe(file_csv, file_json):
...
     # parse strings to np.float64
     is_array = ['Currents', 'Voltages', 'diode_voltage']
-    joined[is_array] = joined[is_array].applymap(
+    joined[is_array] = joined[is_array].map(
         lambda a: np.asarray(a, dtype=np.float64)
     )
     is_number = ['v_oc', 'i_sc', 'v_mp', 'i_mp', 'p_mp', 'i_x', 'i_xx',
                  'Temperature']
-    joined[is_number] = joined[is_number].applymap(np.float64)
+    joined[is_number] = joined[is_number].map(np.float64)
```
Comment (kandersolar) [inline pvlib/tests/test_singlediode.py:118]: Looking at the CI test failures, it seems like `DataFrame.map` only exists in newer versions of pandas.  I think this one can be replaced with just `joined[is_number'].astype(np.float64)`.  For the one on line 113, maybe just fall back to python like this?

```python
    for col in ['Currents', 'Voltages', 'diode_voltage']:
        joined[col] = [np.asarray(a, dtype=np.float64) for a in joined[col]]
```

**Review by kandersolar — COMMENTED** [review]
Thanks @matsuobasho! This is looking pretty good.  There is a linting failure that needs to be addressed, and one pandas compatibility issue mentioned below.

Can you also list yourself as a contributor in the 0.10.3 what's new file?  https://github.com/pvlib/pvlib-python/blob/main/docs/sphinx/source/whatsnew/v0.10.3.rst

**Comment by kandersolar** [comment]
Looking at the [3.11-bare CI logs](https://github.com/pvlib/pvlib-python/actions/runs/6776322189/job/18417944323?pr=1900#step:9:58), this PR gets rid of almost all of the pandas warnings, but there are still a few left.  @matsuobasho do you want to address those as well in this PR to fully close #1892, or should we merge this as-is and fix the rest in another PR?

**Comment by matsuobasho** [comment]
@kandersolar , I'll handle them in this one, since it's in-scope for the issue.  Thanks

**Comment by matsuobasho** [comment]
@kandersolar , can you please provide advice on how to resolve the `invalid value encountered in divide` warning in `tests/test_inverter.py::test_sandia_multi_array`?

Here is the function:
image

Here are the values of `vdcs` and  `idcs`:
image

The warning occurs on l196 of Sandia multi, where we are dividing by 0:
image

Thanks

**Comment by matsuobasho** [comment]
Also, most (if not all) of the other warnings also occur in the underlying functions being called by the test functions.
i.e. `scaling.py`, `snow.py`, `soiling.py`, `solarposition.py`, `bifacial/utils.py`, `singlediode.py`, `ivtools/sdm.py`.

Is fixing these in-scope for this issue?  Technically these aren't the test functions themselves.

**Comment by kandersolar** [comment]
Yes, good point.  I just saw pandas FutureWarnings in the logs but didn't check what files they were coming from :) Let's leave the other warnings for another PR.  

Thanks for this contribution @matsuobasho!

**Review by kandersolar — APPROVED** [review]

**Pull request merged**